routing/payment_lifecycle.go

```diff
@@ -247,7 +247,19 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
 
 		// We found a route to try, launch a new shard.
 		attempt, outcome, err := shardHandler.launchShard(rt)
-		if err != nil {
+		switch {
+		// We can get an already failed error in the case where we
+		// received a permanent failure while we were pathfinding for
+		// this shard. We know that we are in a terminal state, because
+		// we've received this error, so we can continue and wait for
+		// our remaining shards to complete.
+		case err == channeldb.ErrPaymentAlreadyFailed:
```

Comment:
While reviewing I noticed a difference between test vs. production code (pre-dates this PR), which looks like it may have an impact on behavior. I think this solution will cover us in the test, but I have a question as to whether this will cover this bug case "in the wild".

The definition of a "FAILED" payment looks a bit different in testing as compared to the live `resumePayment()` payment lifecycle machine. See the following:
- [Register Attempt (Mock)](https://github.com/lightningnetwork/lnd/blob/master/routing/mock_test.go#L294-L296)
- [Register Attempt (Live)](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294)

Both check to see whether the top level payment is "FAILED" before allowing a new shard registration. The mock determines failure with a simple lookup in the `mockControlTower`'s `failed` map. However in the live case, when we [register a new shard with the ControlTower/DB](https://github.com/lightningnetwork/lnd/blob/master/routing/payment_lifecycle.go#L380), we fetch the payment and [ensure the payment is `IN-FLIGHT`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L288-L294). With the way [`fetchPayment()`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payments.go#L358-L365) works, a payment is reported as `status=IN-FLIGHT` by default if it has _*any in-flight shards*_. Given this, I'm not certain that we would be returning [`ErrPaymentAlreadyFailed`](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L670). If [`shard 1`](https://github.com/lightningnetwork/lnd/pull/5214#issue-615993831) is stuck `IN-FLIGHT`, it looks like `fetchPayment()` will tell us the payment is still in flight.

I think we might be hitting the [terminal info check](https://github.com/lightningnetwork/lnd/blob/v0.12.1-beta.rc6-branch/channeldb/payment_control.go#L300-L302) just below (which exists in the wild but I can't find in our mock). This is a looser check which is sensitive to any payment level failure, but does not care about whether there are in flight shards or not. In this case, it would not be `ErrPaymentAlreadyFailed`, but actually `ErrPaymentTerminal` which we're hitting in the wild.

Let me know if this explanation is difficult to grok or if I'm making up phantom differences which don't actually exist. Here’s a diff with what I think will get the mock of `RegisterAttempt` to be more inline with how it looks like it works in the wild: https://github.com/lightningnetwork/lnd/commit/cb22a03850a78bbfd630c3aa3f4a9d4bb6633971

Reply:
I agree, we should do the terminal info check first, that's the most important condition that must be satisfied.

Good catch on the mocks 👍